content/concepts/peer-routing.md

```diff
@@ -3,6 +3,80 @@ title: Peer Routing
+
+Leaves on the tree that are not peers represent [content](/concepts/content-routing/) which is stored by some close nodes.
+
+![Kademlia-system](Kademlia-system.png)
+
+Red node here has prefix `001`, go from the top to the bottom by edges. Let's say it' Bob. Alice here has prefix `11101` - she is blue node. Yellow leaves are other peers. For any given node, Kademlia divides the whole binary tree into a series of successively lower subtrees that don’t contain the given node. The highest subtree consists of the half of the binary tree not containing the node. The next subtree consists of the half of the remaining tree not containing the node, and so forth. With example it'll be more clear. You can see here for node `001` subtrees are circled and consist of all nodes with prefixes `1`, `01`, `000`.
+
+The Kademlia protocol ensures that every node knows of at least one node in each of its subtrees, if that subtree contains a node. This guarantees that any node can be located by it's peer ID.
+
+Let's see how Bob can locate Alice in this example. Bob knows Alices peerID, but he needs to get the value - Alices IP address. From subtree with prefix `1` Bob knows peer with prefix `100`. He sends a message to their address, where he asks for Alices address.
+
+![1-hop](1-hop.png)
+
+Peer `100` doesn't have Alice's key. But they have key of node `110` from their subtree which locates Alice better! And we are doing second hop:
+
+![2-hop](2-hop.png)
+
+Peer `110` also doesn't have Alice's key, but they again locate Alice better and send request to node with prefix `1111`.
+
+![3-hop](3-hop.png)
+
+Peer `1111` stores Alices peer id and can send the key to the Bob! This is how routing works and here is full scheme:
+
+![final-hop](final-hop.png)
+
+Peer `1111` sends response directly to Bob, since each request in this chain had Bob's address.
+
+### Routing table
+
+The place where one node stores info about other nodes called routing table. The routing table is a binary tree whose leaves are `k-buckets`. Let's boil it down.
+
+What are `k-buckets`? Remember our initial tree? Here it is. ![Kademlia-system](Kademlia-system.png) See this circled sub-trees? Also, remember that Kademlia protocol ensures that every node knows of at least one node in each of its subtrees, if that subtree contains a node. `k-bucket` of a sub-tree is a set of that nodes. Once again: from each subtree we take only those nodes that we know - and we call it `k-bucket`. In out case `k-buckets` for Bob look like this
+
+![k-buckets](k-buckets.png)
+
+Each `k-bucket` covers some range of the ID space, and together the `k-buckets` cover the entire peer ID space with no overlap.
+
+## Dynamic DHT
+
+Okay, that was it for static DHT. And the gap from our current state to a fully dynamic DHT is not as far away as it seems. Let us first consider how we can extend this protocol to support computers sporadically leaving.
```

Comment:
Cool to only introduce this complexity later on in the article :+1:

Reply:
Oh I also like this, it wasn't originally my idea though

I borrowed it from [here](https://codethechange.stanford.edu/guides/guide_kademlia.html)
(and I referred to this source at the end :))